File: src/main/python/AoC2020_22.py
#! /usr/bin/env python3
#
# Advent of Code 2020 Day 22
#

from dataclasses import dataclass
from copy import deepcopy
from aoc import my_aocd
from aoc.common import log


@dataclass
class Players:
    player1: list[int]
    player2: list[int]


def _parse(inputs: tuple[str]) -> Players:
    blocks = my_aocd.to_blocks(inputs)
    pl1 = []
    for i, line in enumerate(blocks[0]):
        if i == 0:
            continue
        pl1.append(int(line))
    pl2 = []
    for i, line in enumerate(blocks[1]):
        if i == 0:
            continue
        pl2.append(int(line))
    return Players(pl1, pl2)


def _get_score(players: Players):
    log("")
    log("")
    log("== Post-game results ==")
    log(f"Player 1's deck: {players.player1}")
    log(f"Player 2's deck: {players.player2}")
    log("")
    log("")
    winner = players.player1 if players.player1 else players.player2
    total = 0
    for i, c in enumerate(winner):
        total += (len(winner)-i) * c
    return total


def _play_combat(players: Players, total_games: int, recursive: bool):
    total_games += 1
    # game = total_games
    # clog(lambda: f"=== Game {game} ===")
    seen = set()
    # rnd = 1
    pl1 = players.player1
    pl2 = players.player2
    while pl1 and pl2:
        round_ = (tuple(pl1), tuple(pl2))
        if recursive and round_ in seen:
            return players, total_games
        seen.add(round_)
        # clog(lambda: "")
        # clog(lambda: f"-- Round {rnd} (Game {game}) --")
        # clog(f"Player 1's deck: {pl1}")
        # clog(lambda: f"Player 2's deck: {pl2}")
        n1 = pl1.pop(0)
        # clog(lambda: f"Player 1 plays: {n1}")
        n2 = pl2.pop(0)
        # clog(lambda: f"Player 2 plays: {n2}")
        if recursive and len(pl1) >= n1 and len(pl2) >= n2:
            # clog(lambda: "Playing a sub-game to determine the winner...")
            # clog(lambda: "")
            pl1_sub = deepcopy(pl1[:n1])
            pl2_sub = deepcopy(pl2[:n2])
            players_sub = Players(pl1_sub, pl2_sub)
            players_sub, total_games = _play_combat(players_sub,
                                                    total_games, True)
            winner = 1 if players_sub.player1 else players_sub.player2
            # clog(lambda: "")
            # clog(lambda: f"...anyway, back to game {game}.")
        else:
            winner = 1 if n1 > n2 else 2
        if winner == 1:
            pl1.append(n1)
            pl1.append(n2)
        else:
            pl2.append(n2)
            pl2.append(n1)
        # clog(lambda: f"Player {winner} wins round {rnd} of game {game}!")
        # if recursive and rnd % 2000 == 0:
        #     print('.', end='', flush=True)
        # rnd += 1
    # clog(lambda: f"The winner of game {game} is player {winner}!")
    return players, total_games


def _play_regular_combat(players: Players, total_games: int):
    return _play_combat(players, total_games, False)


def _play_recursive_combat(players: Players, total_games: int):
    return _play_combat(players, total_games, True)


def part_1(inputs: tuple[str]) -> int:
    players = _parse(inputs)
    players, _ = _play_regular_combat(players, total_games=0)
    return _get_score(players)


def part_2(inputs: tuple[str]) -> int:
    players = _parse(inputs)
    players, _ = _play_recursive_combat(players, total_games=0)
    # print("")
    return _get_score(players)


TEST = """\
Player 1:
9
2
6
3
1

Player 2:
5
8
4
7
10
""".splitlines()


def main() -> None:
    my_aocd.print_header(2020, 22)

    assert part_1(TEST) == 306
    assert part_2(TEST) == 291

    inputs = my_aocd.get_input(2020, 22, 53)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")


if __name__ == '__main__':
    main()

File: src/main/java/AoC2020_22.java
import static java.util.Arrays.asList;
import static java.util.stream.Collectors.toCollection;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import com.github.pareronia.aoc.MutableInt;
import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.StringUtils;
import com.github.pareronia.aoc.solution.Logger;
import com.github.pareronia.aoc.solution.LoggerEnabled;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2020_22
            extends SolutionBase<List<String>, Integer, Integer> {
    
    private AoC2020_22(final boolean debug) {
        super(debug);
    }
    
    public static final AoC2020_22 create() {
        return new AoC2020_22(false);
    }

    public static final AoC2020_22 createDebug() {
        return new AoC2020_22(true);
    }
    
    @Override
    protected List<String> parseInput(final List<String> inputs) {
        return inputs;
    }
    
    @Override
    public Integer solvePart1(final List<String> inputs) {
        final CrabCombat combat = CrabCombat.fromInput(inputs, this.logger);
        combat.playRegular();
        return combat.getScore();
    }
    
    @Override
    public Integer solvePart2(final List<String> inputs) {
        final CrabCombat combat = CrabCombat.fromInput(inputs, this.logger);
        combat.playRecursive();
        return combat.getScore();
    }
    
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "306"),
        @Sample(method = "part2", input = TEST, expected = "291"),
        @Sample(method = "part2", input = LOOP, expected = "105"),
    })
    public static void main(final String[] args) throws Exception {
        AoC2020_22.create().run();
    }
    
    private static final String TEST = """
        Player 1:
        9
        2
        6
        3
        1
        
        Player 2:
        5
        8
        4
        7
        10
        """;
    
    private static final String LOOP = """
        Player 1:
        43
        19
        
        Player 2r
        2
        29
        14
        """;
    
    record CrabCombat(
            List<Integer> player1,
            List<Integer> player2,
            Logger logger
    ) implements LoggerEnabled {
        
        public static CrabCombat fromInput(
                final List<String> inputs, final Logger logger
        ) {
            final List<List<String>> blocks = StringOps.toBlocks(inputs);
            final List<Integer> player1 = new ArrayList<>();
            final List<Integer> player2 = new ArrayList<>();
            blocks.get(0).stream().skip(1)
                    .map(Integer::valueOf)
                    .collect(toCollection(() -> player1));
            blocks.get(1).stream().skip(1)
                    .map(Integer::valueOf)
                    .collect(toCollection(() -> player2));
            return new CrabCombat(player1, player2, logger);
        }

        @Override
        public Logger getLogger() {
            return logger;
        }

        public void playRegular() {
            play(this, new MutableInt(), false);
        }
        
        public void playRecursive() {
            play(this, new MutableInt(), true);
        }
        
        private void play(
                final CrabCombat combat,
                final MutableInt _game,
                final boolean recursive
        ) {
            _game.increment();
            final int game = _game.intValue();
            log(() -> String.format("=== Game %d ===", game));
            final Set<Round> seen = new HashSet<>();
            final MutableInt rnd = new MutableInt(1);
            final List<Integer> pl1 = combat.player1;
            final List<Integer> pl2 = combat.player2;
            while (!pl1.isEmpty() && !pl2.isEmpty()) {
                log(() -> "");
                log(() -> String.format("-- Round %d (Game %d) --", rnd.intValue(), game));
                log(() -> String.format("Player 1's deck: %s", StringUtils.join(pl1, ", ")));
                log(() -> String.format("Player 2's deck: %s", StringUtils.join(pl2, ", ")));
                final Round round = new Round(new ArrayList<>(pl1), new ArrayList<>(pl2));
                if (recursive && seen.contains(round)) {
                    pl2.clear();
                    break;
                }
                seen.add(round);
                final Integer n1 = pl1.remove(0);
                log(() -> String.format("Player 1 plays: %d", n1));
                final Integer n2 = pl2.remove(0);
                log(() -> String.format("Player 2 plays: %d", n2));
                final Integer winner;
                if (recursive && pl1.size() >= n1 && pl2.size() >= n2) {
                    log(() -> "Playing a sub-game to determine the winner...");
                    log(() -> "");
                    final List<Integer> pl1_sub = new ArrayList<>(pl1.subList(0, n1));
                    final List<Integer> pl2_sub = new ArrayList<>(pl2.subList(0, n2));
                    play(new CrabCombat(pl1_sub, pl2_sub, logger), _game, true);
                    log(() -> "");
                    log(() -> String.format("...anyway, back to game %d.", game));
                    winner = pl2_sub.isEmpty() ? 1 : 2;
                } else {
                    winner = n1 > n2 ? 1 : 2;
                }
                if (winner == 1) {
                    pl1.addAll(asList(n1, n2));
                } else {
                    pl2.addAll(asList(n2, n1));
                }
                log(() -> String.format("Player %d wins round %d of game %d!", winner, rnd.intValue(), game));
                rnd.increment();
            }
            final Integer winner = pl2.isEmpty() ? 1 : 2;
            log(() -> String.format("The winner of game %d is player %d!", game, winner));
        }
        
        public int getScore() {
            log("");
            log("");
            log("== Post-game results ==");
            log(String.format("Player 1's deck: %s", StringUtils.join(player1, ", ")));
            log(String.format("Player 2's deck: %s", StringUtils.join(player2, ", ")));
            log("");
            log("");
            final List<Integer> winner = player2.isEmpty() ? player1 : player2;
            int total = 0;
            for (int i = 0; i < winner.size(); i++) {
                total += (winner.size() - i) * winner.get(i);
            }
            return total;
        }
        
        record Round(List<Integer> pl1, List<Integer> pl2) {}
    }
}
